statsmodels/tsa/vecm/vecm.py

```diff
@@ -0,0 +1,2146 @@
+    eig = np.linalg.eig(chain_dot(s01_s11_.T, inv(s00), s01_s11_))
+    lambd = eig[0]
+    v = eig[1]
+    # reorder eig_vals to make them decreasing (and order eig_vecs accordingly)
+    lambd_order = np.argsort(lambd)[::-1]
+    lambd = lambd[lambd_order]
+    v = v[:, lambd_order]
+    return s00, s01, s10, s11, s11_, lambd, v
+
+
+def select_coint_rank(endog_tot, det_order, k_ar, method="trace", signif=0.95,
+                      verbose=True):
+    """
+    Calculate the cointegration rank of a VECM.
+
+    Parameters
+    ----------
+    endog_tot : array-like (nobs_tot x neqs)
+        The data with presample.
+    det_order : int
+        * -1 - no deterministic terms
+        * 0 - constant term
+        * 1 - linear trend
+        * >1 - higher polynomial order
+    k_ar : int, nonnegative
+        Number of lagged differences in the model.
+    method : str, {"trace", "maxeig"}, default: "trace"
+        If "trace", the trace test statistic is used. If "maxeig", the
+        maximum eigenvalue test statistic is used.
+    signif : float, {0.9, 0.95, 0.99}, default: 0.95
+        The test's significance level.
+
+    Returns
+    -------
+    rank : int
+        The cointegration rank suggested by the test.
+    """
+
+    if method not in ["trace", "maxeig"]:
+        raise ValueError("The method argument has to be either 'trace' or"
+                         "'maximum eigenvalue'.")
+
+    possible_signif_values = [0.9, 0.95, 0.99]
+    if signif not in possible_signif_values:
+        raise ValueError("Please choose a significance level from {0.9, 0.95,"
+                         "0.99}")
+
+    coint_result = coint_johansen(endog_tot, det_order, k_ar)
+    neqs = endog_tot.shape[1]
+    r_0 = 0  # rank in null hypothesis
+    test_stat = coint_result.lr1 if method == "trace" else coint_result.lr2
+    crit_vals = coint_result.cvt if method == "trace" else coint_result.cvm
+    signif_index = possible_signif_values.index(signif)
+
+    verbose_data = []
+    while r_0 < neqs:
+        r_1 = neqs if method == "trace" else (r_0 + 1)
+        test_statistic = test_stat[r_0]
+        critical_value = crit_vals[r_0, signif_index]
+
+        if verbose:
+            verbose_data.append([r_0, r_1, round(test_statistic, 4),
+                                 critical_value])
+
+        if test_statistic < critical_value:
+            break  # we accept current rank
+        else:
+            r_0 += 1  # we reject current rank and test next possible rank
+
+    if verbose:
+        headers = ["r_0", "r_1", "test statistic", "critical value"]
+        title = "Johansen cointegration test using " + \
+                ("trace" if method == "trace" else "maximum eigenvalue") + \
+                " test statistic with " + str(round(signif*100)) + \
+                "% significance level"
+        print(SimpleTable(data=verbose_data, headers=headers, title=title))
+
+    return r_0
+
+
+def coint_johansen(endog_tot, det_order, k_ar):
```

Comment:
I have basically moved your function to this place in order to have all VECM-related functionality inside the vecm-module. Maybe we should add an underscore to mark it as internal because the user will probably want to use `select_coint_rank` as this function is calling `coint_johansen` until the optimal rank has been found. What do you think?